### PR TITLE
Port CreateAcceptance to the TCK

### PR DESCRIPTION
**Pull request by petraselmer**
Only merge this PR once https://github.com/neo4j/neo4j/pull/7215 has been merged

### CODE REVIEW

File: tck/features/CreateAcceptance.feature [hunk 1]
```diff
@@ -0,0 +1,607 @@
...
+    And having executed:
+      """
+      MATCH (x:X), (y:Y)
+      CREATE (x)<-[:TYPE]-(y)
+      """
+    When executing query:
+      """
+      MATCH (x:X)<-[:TYPE]-(y:Y)
+      RETURN x, y
+      """
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: I find that we want to have the query under test in the `When` step, and given that this file is meant to test `CREATE`, I'd say we want a query with `CREATE` in the `When` steps. It seems the `And` step has the create query here, and that the `When` step is meant to make assertions on the result. In #54 I added another `And` step that is meant to be run after the `When` step, to make possible such extra, query-based assertions after an updating query (I call these 'control queries').

Perhaps that's a better approach to this (and subsequent) scenarios? I do like to retain that `When` contains the query that the scenario wants to test.

File: tck/features/CreateAcceptance.feature [hunk 2]
```diff
@@ -0,0 +1,607 @@
...
+    And having executed:
+      """
+      CREATE (:Begin)
+      """
+    And having executed:
+      """
+      MATCH (x:Begin)
+      CREATE (x)-[:TYPE]->(:End)
+      """
+    When executing query:
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: See long comment above.

File: tck/features/CreateAcceptance.feature [hunk 3]
```diff
@@ -0,0 +1,607 @@
...
+      | +relationships | 1 |
+      | +labels        | 2 |
+
+  Scenario: Create a relationship with a reversed direction - 2
+    Given any graph
+    And having executed:
+      """
+      CREATE (:A)<-[:R]-(:B)
+      """
+    When executing query:
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: See long comment above.

File: tck/features/CreateAcceptance.feature [hunk 4]
```diff
@@ -0,0 +1,607 @@
...
+      | +relationships | 2 |
+      | +labels        | 3 |
+
+  Scenario: Create a pattern with multiple hops - 2
+    Given any graph
+    And having executed:
+      """
+      CREATE (:A)-[:R]->(:B)-[:R]->(:C)
+      """
+    When executing query:
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: See long comment above.

File: tck/features/CreateAcceptance.feature [hunk 5]
```diff
@@ -0,0 +1,607 @@
...
+      """
+      MATCH ()
+      CREATE ()
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +nodes  | 1 |
+
+  Scenario: Create a single node with properties
+    Given any graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: I don't think this is true. In the case of a populated uniqueness constraint, the query will not behave as specified here. I think we need `Given empty graph` here.

File: tck/features/CreateAcceptance.feature [hunk 6]
```diff
@@ -0,0 +1,607 @@
...
+      """
+    Then the result should be:
+      | p     |
+      | 'foo' |
+    And the side effects should be:
+      | +nodes      | 1 |
+      | +properties | 1 |
+
+  Scenario: Creating a node with null properties should not return those properties
+    Given any graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: As above, `Given empty graph`.

File: tck/features/CreateAcceptance.feature [hunk 7]
```diff
@@ -0,0 +1,607 @@
...
+      """
+    Then the result should be:
+      | id |
+      | 12 |
+    And the side effects should be:
+      | +nodes      | 1 |
+      | +properties | 1 |
+
+  Scenario: Creating a relationship with null properties should not return those properties
+    Given any graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: As above, not applicable on any graph.

File: tck/features/CreateAcceptance.feature [hunk 8]
```diff
@@ -0,0 +1,607 @@
...
+    Then the result should be:
+      | r.id |
+      | 12   |
+    And the side effects should be:
+      | +nodes         | 2 |
+      | +relationships | 1 |
+      | +properties    | 1 |
+
+  Scenario: Create a simple pattern
+    Given any graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: empty

File: tck/features/CreateAcceptance.feature [hunk 9]
```diff
@@ -0,0 +1,607 @@
...
+      """
+      CREATE ()-[:R]->()
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +nodes         | 2 |
+      | +relationships | 1 |
+
+  Scenario: Create a self loop
+    Given any graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: empty

File: tck/features/CreateAcceptance.feature [hunk 10]
```diff
@@ -0,0 +1,607 @@
...
+      CREATE (root:R)-[:LINK]->(root)
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +nodes         | 1 |
+      | +relationships | 1 |
+      | +labels        | 1 |
+
+  Scenario: Create a self loop using MATCH
+    Given any graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: And the last case, when there's a node property existence constraint on label `:R`, this won't work.

File: tck/features/CreateAcceptance.feature [hunk 11]
```diff
@@ -0,0 +1,607 @@
...
+      | +relationships | 2 |
+      | +labels        | 3 |
+
+  Scenario: Create a pattern with multiple hops in the reverse direction - 2
+    Given any graph
+    And having executed:
+      """
+      CREATE (:A)<-[:R]-(:B)<-[:R]-(:C)
+      """
+    When executing query:
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: See long comment above.

File: tck/features/CreateAcceptance.feature [hunk 12]
```diff
@@ -0,0 +1,607 @@
...
+      | +relationships | 2 |
+      | +labels        | 3 |
+
+  Scenario: Create a pattern with multiple hops in varying directions - 2
+    Given any graph
+    And having executed:
+      """
+      CREATE (:A)-[:R]->(:B)<-[:R]-(:C)
+      """
+    When executing query:
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: See long comment above.

File: tck/features/CreateAcceptance.feature [hunk 13]
```diff
@@ -0,0 +1,607 @@
...
+      | +nodes         | 4 |
+      | +relationships | 3 |
+
+  Scenario: Create a pattern with multiple hops with multiple types and varying directions - 2
+    Given any graph
+    And having executed:
+      """
+      CREATE ()-[:R1]->()<-[:R2]-()-[:R3]->()
+      """
+    When executing query:
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: See long comment above.

File: tck/features/LargeCreateQuery.feature [hunk 14]
```diff
@@ -0,0 +1,1360 @@
...
+    When executing query:
+      """
+      CREATE (TheMatrix:Movie {title: 'The Matrix', released: 1999, tagline: 'Welcome to the Real World'})
+      CREATE (Keanu:Person {name: 'Keanu Reeves', born: 1964})
+      CREATE (Carrie:Person {name: 'Carrie-Anne Moss', born: 1967})
+      CREATE (Laurence:Person {name: 'Laurence Fishburne', born: 1961})
+      CREATE (Hugo:Person {name: 'Hugo Weaving', born: 1960})
+      CREATE (AndyW:Person {name: 'Andy Wachowski', born: 1967})
+      CREATE (LanaW:Person {name: 'Lana Wachowski', born: 1965})
+      CREATE (JoelS:Person {name: 'Joel Silver', born: 1952})
```
Comment (Mats-SX) [inline tck/features/LargeCreateQuery.feature]: These variables aren't following Cypher style, as they are not camel case'd.

File: tck/features/LargeCreateQuery.feature [hunk 15]
```diff
@@ -0,0 +1,1360 @@
...
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +nodes         | 171 |
+      | +relationships | 253 |
+      | +properties    | 564 |
+      | +labels        | 171 |
+
+  Scenario: Many CREATE clauses
+    Given any graph
```
Comment (Mats-SX) [inline tck/features/LargeCreateQuery.feature]: empty

File: tck/features/LargeCreateQuery.feature [hunk 16]
```diff
@@ -0,0 +1,1360 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+Feature: LargeCreateQuery
+
+  Scenario: Generate the movie graph correctly
+    Given any graph
```
Comment (Mats-SX) [inline tck/features/LargeCreateQuery.feature]: empty

**Comment by Mats-SX** [comment]
@petraselmer I stopped putting down line comments for the `any graph` vs `empty graph` condition. I think most of the scenarios need `empty graph`. I hadn't realised that creating a labelless node should always be possible, though, so that's good to know. But usually, when doing something that accesses the graph, `Given any graph` is too loose a condition to be able to state the outcome.

File: tck/features/CreateAcceptance.feature (hunk 1 above)
Comment (petraselmer) [inline tck/features/CreateAcceptance.feature]: Totally agreed - will redo all to follow this format

File: tck/features/CreateAcceptance.feature [hunk 17]
```diff
@@ -0,0 +1,590 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+Feature: CreateAcceptance
+
+  Scenario: Create a single node
+    Given an empty graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: This could be `any graph`

File: tck/features/CreateAcceptance.feature [hunk 18]
```diff
@@ -0,0 +1,590 @@
...
+    When executing query:
+      """
+      CREATE ()
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +nodes  | 1 |
+
+  Scenario: Create a single node with a single label
+    Given an empty graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: This could be `any graph`

File: tck/features/CreateAcceptance.feature [hunk 19]
```diff
@@ -0,0 +1,590 @@
...
+      """
+      CREATE (:A)
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +nodes  | 1 |
+      | +labels | 1 |
+
+  Scenario: Create a single node with multiple labels
+    Given an empty graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: This could be `any graph`

File: tck/features/CreateAcceptance.feature [hunk 20]
```diff
@@ -0,0 +1,590 @@
...
+      """
+      MATCH ()
+      CREATE ()
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +nodes  | 1 |
+
+  Scenario: Create a single node with properties
+    Given an empty graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: This could be `any graph`

File: tck/features/CreateAcceptance.feature [hunk 21]
```diff
@@ -0,0 +1,590 @@
...
+      | 'foo' |
+    And the side effects should be:
+      | +nodes      | 1 |
+      | +properties | 1 |
+
+  Scenario: Creating a node with null properties should not return those properties
+    Given an empty graph
+    When executing query:
+      """
+      CREATE (n {id: 12, property: NULL})
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: `NULL` -> `null`

File: tck/features/CreateAcceptance.feature [hunk 22]
```diff
@@ -0,0 +1,590 @@
...
+      """
+    Then the result should be:
+      | p     |
+      | 'foo' |
+    And the side effects should be:
+      | +nodes      | 1 |
+      | +properties | 1 |
+
+  Scenario: Creating a node with null properties should not return those properties
+    Given an empty graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: This could be `any graph`

File: tck/features/CreateAcceptance.feature [hunk 23]
```diff
@@ -0,0 +1,590 @@
...
+      """
+    Then the result should be:
+      | id |
+      | 12 |
+    And the side effects should be:
+      | +nodes      | 1 |
+      | +properties | 1 |
+
+  Scenario: Creating a relationship with null properties should not return those properties
+    Given an empty graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: This could be `any graph`

File: tck/features/CreateAcceptance.feature [hunk 24]
```diff
@@ -0,0 +1,590 @@
...
+    When executing query:
+      """
+      CREATE ()-[r:X {id: 12, property: NULL}]->()
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: `NULL` -> `null`

File: tck/features/CreateAcceptance.feature [hunk 25]
```diff
@@ -0,0 +1,590 @@
...
+    Then the result should be:
+      | r.id |
+      | 12   |
+    And the side effects should be:
+      | +nodes         | 2 |
+      | +relationships | 1 |
+      | +properties    | 1 |
+
+  Scenario: Create a simple pattern
+    Given an empty graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: This could be `any graph`

File: tck/features/CreateAcceptance.feature [hunk 26]
```diff
@@ -0,0 +1,590 @@
...
+      """
+      CREATE ()-[:R]->()
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +nodes         | 2 |
+      | +relationships | 1 |
+
+  Scenario: Create a self loop
+    Given an empty graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: This could be `any graph`

File: tck/features/CreateAcceptance.feature [hunk 27]
```diff
@@ -0,0 +1,590 @@
...
+      """
+      MATCH (root:R)
+      CREATE (root)-[:LINK]->(root)
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +relationships | 1 |
+
+  Scenario: Create nodes and relationships
+    Given an empty graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: This could be `any graph`

File: tck/features/CreateAcceptance.feature [hunk 28]
```diff
@@ -0,0 +1,590 @@
...
+      CREATE (a), (b),
+             (a)-[:R]->(b)
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +nodes         | 2 |
+      | +relationships | 1 |
+
+  Scenario: Create a relationship with a property
+    Given an empty graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: This could be `any graph`

File: tck/features/CreateAcceptance.feature [hunk 29]
```diff
@@ -0,0 +1,590 @@
...
+      """
+      CREATE (:X)
+      CREATE (:Y)
+      """
+    When executing query:
+      """
+      MATCH (x:X), (y:Y)
+      CREATE (x)<-[:TYPE]-(y)
+      """
+    And having executed:
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: This should be `When executing control query`

File: tck/features/CreateAcceptance.feature [hunk 30]
```diff
@@ -0,0 +1,590 @@
...
+      | +relationships | 1 |
+
+  Scenario: Create a relationship with the correct direction - 2
+    Given an empty graph
+    And having executed:
+      """
+      CREATE (:X)
+      CREATE (:Y)
+      """
+    When executing query:
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: This step should have its own `Then the result should be empty` and `And the side effects should be`

File: tck/features/CreateAcceptance.feature [hunk 31]
```diff
@@ -0,0 +1,590 @@
...
+      CREATE ()
+      WITH *
+      CREATE ()
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +nodes | 4 |
+
+  Scenario: Create a relationship with a reversed direction - 1
+    Given an empty graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: This could be `any graph`

File: tck/features/CreateAcceptance.feature [hunk 32]
```diff
@@ -0,0 +1,590 @@
...
+      MATCH (a:A)<-[:R]-(b:B)
+      RETURN a, b
+      """
+    Then the result should be:
+      | a    | b    |
+      | (:A) | (:B) |
+    And no side effects
+
+  Scenario: Create a pattern with multiple hops - 1
+    Given an empty graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: This could be `any graph`

File: tck/features/CreateAcceptance.feature [hunk 33]
```diff
@@ -0,0 +1,590 @@
...
+      MATCH (a:A)-[:R]->(b:B)-[:R]->(c:C)
+      RETURN a, b, c
+      """
+    Then the result should be:
+      | a    | b    | c    |
+      | (:A) | (:B) | (:C) |
+    And no side effects
+
+  Scenario: Create a pattern with multiple hops in the reverse direction - 1
+    Given an empty graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: This could be `any graph`

File: tck/features/CreateAcceptance.feature [hunk 34]
```diff
@@ -0,0 +1,590 @@
...
+      MATCH (a)<-[:R]-(b)<-[:R]-(c)
+      RETURN a, b, c
+      """
+    Then the result should be:
+      | a    | b    | c    |
+      | (:A) | (:B) | (:C) |
+    And no side effects
+
+  Scenario: Create a pattern with multiple hops in varying directions - 1
+    Given an empty graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: This could be `any graph`

File: tck/features/CreateAcceptance.feature [hunk 35]
```diff
@@ -0,0 +1,590 @@
...
+      MATCH (a:A)-[r1:R]->(b:B)<-[r2:R]-(c:C)
+      RETURN a, b, c
+      """
+    Then the result should be:
+      | a    | b    | c    |
+      | (:A) | (:B) | (:C) |
+    And no side effects
+
+  Scenario: Create a pattern with multiple hops with multiple types and varying directions - 1
+    Given an empty graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: This could be `any graph`

File: tck/features/CreateAcceptance.feature [hunk 36]
```diff
@@ -0,0 +1,590 @@
...
+      """
+    Then the result should be:
+      | x |
+      | 0 |
+    And the side effects should be:
+      | +nodes         | 2 |
+      | +relationships | 2 |
+
+  Scenario: A bound node should be recognized after projection with WITH + WITH
+    Given an empty graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: This could be `any graph`

File: tck/features/CreateAcceptance.feature [hunk 37]
```diff
@@ -0,0 +1,590 @@
...
+      CREATE (b)
+      CREATE (a)<-[:T]-(b)
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +nodes         | 2 |
+      | +relationships | 1 |
+
+  Scenario: A bound node should be recognized after projection with WITH + UNWIND
+    Given an empty graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: This could be `any graph`

File: tck/features/CreateAcceptance.feature [hunk 38]
```diff
@@ -0,0 +1,590 @@
...
+      CREATE (a)<-[:T]-(b)
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +nodes         | 4 |
+      | +relationships | 2 |
+
+
+  Scenario: Fail when trying to create using an undirected relationship pattern
+    Given an empty graph
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: This could be `any graph`

File: tck/features/LargeCreateQuery.feature [hunk 39]
```diff
@@ -0,0 +1,1361 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+Feature: LargeCreateQuery
+
+  Scenario: Generate the movie graph correctly
+    Given an empty graph
```
Comment (Mats-SX) [inline tck/features/LargeCreateQuery.feature]: I _think_ this could be `any graph`.

File: tck/features/LargeCreateQuery.feature [hunk 40]
```diff
@@ -0,0 +1,1361 @@
...
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +nodes         | 171 |
+      | +relationships | 253 |
+      | +properties    | 564 |
+      | +labels        | 171 |
+
+  Scenario: Many CREATE clauses
+    Given an empty graph
```
Comment (Mats-SX) [inline tck/features/LargeCreateQuery.feature]: I _think_ this could be `any graph`.

**Comment by Mats-SX** [comment]
@petraselmer I'm sorry for my above comment, and I don't agree with it anymore, for the most part. Now that we've cut the scope of openCypher to exclude constraints, creating is always possible. And if the scenario does not contain a control query with `MATCH`, then `any graph` is preferable.

Also, it seems we are not agreeing fully on the design of the control queries. You seem to have multiple scenarios to first assert on the `When` query, and then to assert on the control query. I prefer that we do this in one scenario, as per https://github.com/opencypher/openCypher/blob/master/tck/features/MergeIntoAcceptance.feature#L27. I think it makes sense to contain the assertion of the semantics in a single scenario, instead of splitting it up.

**Comment by petraselmer** [comment]
@Mats-SX I have (hopefully) addressed your comments. No probs about the `any graph`. Also, the 'control query' construct is great - I didn't know about this before - and this gets rid of the frankly clumsy multiple scenario set up. I have now done this across the board...

File: tck/features/CreateAcceptance.feature [hunk 41]
```diff
@@ -0,0 +1,523 @@
...
+      """
+      CREATE ()
+      """
+    When executing query:
+      """
+      MATCH (n)
+      MATCH (m)
+      WITH n AS a, m AS b
+      CREATE (a)-[:T]->(b)
+      RETURN id(a) AS a, id(b) AS b
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: `id()` function not allowed; use a property instead.

File: tck/features/CreateAcceptance.feature [hunk 42]
```diff
@@ -0,0 +1,523 @@
...
+      """
+    When executing query:
+      """
+      MATCH (n)
+      MATCH (m)
+      WITH n AS a, m AS b
+      CREATE (a)-[:T]->(b)
+      WITH a AS x, b AS y
+      CREATE (x)-[:T]->(y)
+      RETURN id(x) AS x, id(y) AS y
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: `id()` function not allowed; use a property instead.

File: tck/features/CreateAcceptance.feature [hunk 43]
```diff
@@ -0,0 +1,523 @@
...
+      CREATE ()
+      """
+    When executing query:
+      """
+      MATCH (n)
+      WITH n AS a
+      CREATE (a)-[:T]->()
+      WITH a AS x
+      CREATE (x)-[:T]->()
+      RETURN id(x) AS x
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: `id()` function not allowed; use a property instead.

File: tck/features/CreateAcceptance.feature [hunk 44]
```diff
@@ -0,0 +1,523 @@
...
+    And having executed:
+      """
+      CREATE ()
+      """
+    When executing query:
+      """
+      MATCH (n)
+      WITH n AS a
+      CREATE (a)-[:T]->()
+      RETURN id(a) AS a
```
Comment (Mats-SX) [inline tck/features/CreateAcceptance.feature]: `id()` function not allowed; use a property instead.

**Comment by Mats-SX** [comment]
@petraselmer Except for the `id()` function, this looks good. I find that usually the easiest way is to simply create an `id` property on the nodes, but it's also possible to work around by labelling and projecting the full nodes, or their labels using `labels()`.

**Pull request merged**